text/0026-transform_application.md

```diff
@@ -0,0 +1,107 @@
+
+Getting down to it, there's a fundamental mismatch with how transforms are defined, and how FSI assumes it may operate. Before FSI, there was a one-to-one correlation between running a transform and creating a commit, hence why it made sense to tie that execution to the "save" command. The `--dry-run` flag was acting as a development tool. But with FSI, that original assumption no longer applies. A transform should be able to be executed as a way to modify the body without committing, in fact this is probably going to be the expected behavior for any new user coming to qri.
+
+# Proposal
+
+Transforms should not be thought of as equivalent to a state transition to a new commit. They should be viewed literally as a *transform*, from some input to some output, and only sometimes is that performed as creating a commit.
+
+Let's say we add a command `qri apply`, which runs a transform to arrive at some output; but how that output is applies depends upon the context of the command.
+
+1) When operated in the context of a manual save, `apply` can write to stdout, replacing the need for `--dry-run`. In this case, we know that no commit is written, so we can change `save` to write to stdout in other cases.
+2) A transform as a state transition can be thought of as a `qri apply` operation piped into the `qri save` command, with appropriate syntacic sugar.
+3) For FSI, `qri apply` can modify the body in the working directory, letting the user see the results without messing with the `save` and `--dry-run` comamnds. When they eventually do run `qri save`, we should stash the current body file, apply the transform and verify one of the following:
+  * The stashed body matches the output of the transform
+  * The stashed body is unchanged from the previous commit, replace it with the transforms' output
+  * Otherwise throw an error because the user may be making a mistake
+4) Not mentioned above, but once Desktop has better support for transform, we probably will eventually want some interactive development environment. Perhaps transforms can be incrementally compiled and applied, resulting in fast turnaround.
+
+To expand on this idea, applying a transform could be generalized out to other tasks that execute some sort of data pipeline. For example, if we bring back viz, generating those visuals could also be seen as a type of apply operation, with the script being some sort of imaging library, and the output being static images. Or, our `qri update` feature could be somehow incorporated into this. Any sort of task that computes using a dataset to produce some view or artifact could be folded into this command. Better to have a generalized tool than create a one-off command that solves only a narrow problem.
+
+More details of the implementation should wait until the feature is further discussed and some rough consensus can be reached.
```

Comment:
I think we should continue to rely on file extensions for syntax, and magic filenames for FSI integrations. 


What if `qri save` got an `apply` flag? Working outside of FSI that could look like this:

```
qri save --body body.csv --apply transform.star me/dataset
```

in FSI land a file named `transform.star` would be auto-applied. I could also choose to _not_ have a file called `transform.star` and instead  have lots of other scripts in my FSI-linked directory that I could override with `--apply`. `qri save --apply correct_my_math.star` feels pretty natural